File: scripts/cli/triage.mjs
// `cli.mjs triage` — publish/read the nightly-triage DIGEST without hand-editing
// the file. Lazily loaded by ../cli.mjs. The nightly triage agent (a scheduled
// Claude session, see task #35) reads the board read-only via `todos list --json`,
// reasons about it, and writes its findings here with `triage publish`. The
// tracker watches the resulting `triage-digest.json` and raises a desktop
// notification + an in-app summary (later phases).
//
// Why a separate file + CLI (not todos.json): the digest is DERIVED, regenerated
// each run, and strictly advisory — it must never touch the user's board (the
// triage agent is read-only over todos by design). Keeping it in its own file
// means a bad/partial digest can't corrupt the shared todo list, and the writer
// path is one validated, atomic temp+rename — mirroring todos.mjs / todos.rs.
//
// Commands (run as `cli.mjs triage <cmd>`):
//   publish [--file <path>] [--json <inline>]   write the digest (else read stdin)
//   show [--json]                               print the current digest
//   clear                                       remove the digest file
//
// The on-disk shape (snake_case, matching the rest of the app's wire format),
// mirrored by src-tauri/src/triage.rs::TriageDigest:
//   {
//     "version": 1,
//     "generated_at": "<ISO-8601>",      // stamped by publish if absent
//     "project": "<basename>" | null,    // which board this triage covered
//     "headline": "<short line>",        // <=140 chars, for the notification
//     "summary": "<prose>",              // human digest, for the in-app card
//     "items": [                         // findings + suggestions
//       { "kind": "stale|overdue|no_priority|suggestion",
//         "number": <int>|null, "id": "<uuid>"|null,
//         "subject": "<str>", "note": "<str>" }
//     ]
//   }
//
// Exit code is non-zero on any error (bad JSON, bad shape, usage), so the caller
// (and the triage prompt) can tell success from failure.

import { readFileSync, writeFileSync, renameSync, unlinkSync } from "node:fs";
import path from "node:path";

// Finding kinds the digest understands. Keep in lockstep with triage.rs::KINDS
// and the in-app digest view. `stale`/`overdue`/`no_priority` are facts the agent
// surfaces; `suggestion` is an advisory move it proposes (never applied here).
const KINDS = ["stale", "overdue", "no_priority", "suggestion"];

// Same app data dir the tracker, todos CLI, and hook use; the digest lives next
// to todos.json so the tracker finds it without extra config.
function digestPath() {
  const appData =
    process.env.APPDATA ||
    path.join(process.env.USERPROFILE || "", "AppData", "Roaming");
  return path.join(
    appData,
    "com.claude-usage-tracker.app",
    "triage-digest.json",
  );
}

function fail(msg) {
  process.stderr.write(msg + "\n");
  process.exit(1);
}

// Atomic write: serialize to a sibling temp file, then rename over the target
// (rename replaces the destination on Windows). 2-space pretty-print matches the
// tracker's serde output so the file stays hand-readable.
function save(file, data) {
  const tmp = file + ".tmp";
  writeFileSync(tmp, JSON.stringify(data, null, 2) + "\n");
  renameSync(tmp, file);
}

// Forgiving read: a missing/corrupt file yields null (no digest yet), mirroring
// the read-side contract in triage.rs::load. `show` turns that into a friendly
// note; the tracker treats it as "nothing to surface".
function load(file) {
  try {
    const data = JSON.parse(readFileSync(file, "utf8"));
    return data && typeof data === "object" ? data : null;
  } catch {
    return null;
  }
}

// Minimal `--flag value` parser (same shape as todos.mjs): positional args plus
// flag pairs; a flag with no following value becomes `true`.
function parseArgs(args) {
  const flags = {};
  const positional = [];
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a.startsWith("--")) {
      const next = args[i + 1];
      if (next === undefined || next.startsWith("--")) {
        flags[a.slice(2)] = true;
      } else {
        flags[a.slice(2)] = next;
        i++;
      }
    } else {
      positional.push(a);
    }
  }
  return { positional, flags };
}

const PUBLISH_USAGE =
  "usage: cli triage publish [--file <path>] [--json <inline>]\n" +
  "       (with neither flag, the digest JSON is read from stdin)";

// Read the raw digest JSON the caller is publishing: --file <path>, --json
// <inline>, or stdin (fd 0). Exactly one source; --file wins over --json.
function readInput(flags) {
  if (typeof flags.file === "string") {
    try {
      return readFileSync(flags.file, "utf8");
    } catch (e) {
      fail(`cannot read --file ${flags.file}: ${(e && e.message) || e}`);
    }
  }
  if (typeof flags.json === "string") return flags.json;
  try {
    return readFileSync(0, "utf8");
  } catch {
    fail(PUBLISH_USAGE);
  }
}

// Validate + normalize the incoming digest, then write it atomically. Unlike the
// forgiving read side, publish is STRICT: a malformed digest is rejected (exit 1)
// rather than written, so the tracker never has to defend against junk and the
// triage prompt gets a clear failure it can retry. Stamps version=1 and, if the
// caller didn't supply one, generated_at=now.
function cmdPublish(args) {
  const { flags } = parseArgs(args);
  const raw = readInput(flags);

  let input;
  try {
    input = JSON.parse(raw);
  } catch (e) {
    fail(`digest is not valid JSON: ${(e && e.message) || e}`);
  }
  if (!input || typeof input !== "object" || Array.isArray(input)) {
    fail("digest must be a JSON object");
  }

  // Optional scalar fields — present → must be the right type.
  for (const key of ["headline", "summary"]) {
    if (input[key] != null && typeof input[key] !== "string") {
      fail(`"${key}" must be a string`);
    }
  }
  if (
    input.project != null &&
    typeof input.project !== "string"
  ) {
    fail('"project" must be a string or null');
  }
  if (
    input.generated_at != null &&
    typeof input.generated_at !== "string"
  ) {
    fail('"generated_at" must be an ISO-8601 string');
  }

  // Items: an array of findings. Each needs a known kind and a non-empty subject;
  // number/id/note are optional but type-checked when present.
  const rawItems = input.items == null ? [] : input.items;
  if (!Array.isArray(rawItems)) fail('"items" must be an array');
  const items = rawItems.map((it, i) => {
    if (!it || typeof it !== "object" || Array.isArray(it))
      fail(`items[${i}] must be an object`);
    if (!KINDS.includes(it.kind))
      fail(`items[${i}].kind must be one of: ${KINDS.join(" | ")}`);
    if (typeof it.subject !== "string" || !it.subject.trim())
      fail(`items[${i}].subject must be a non-empty string`);
    if (it.number != null && !Number.isInteger(it.number))
      fail(`items[${i}].number must be an integer`);
    if (it.id != null && typeof it.id !== "string")
      fail(`items[${i}].id must be a string`);
    if (it.note != null && typeof it.note !== "string")
      fail(`items[${i}].note must be a string`);
    return {
      kind: it.kind,
      number: it.number == null ? null : it.number,
      id: it.id == null ? null : it.id,
      subject: it.subject,
      note: typeof it.note === "string" ? it.note : "",
    };
  });

  const digest = {
    version: 1,
    generated_at:
      typeof input.generated_at === "string"
        ? input.generated_at
        : new Date().toISOString(),
    project: typeof input.project === "string" ? input.project : null,
    headline: typeof input.headline === "string" ? input.headline : "",
    summary: typeof input.summary === "string" ? input.summary : "",
    items,
  };

  const file = digestPath();
  save(file, digest);
  process.stdout.write(
    `ok: published digest (${items.length} item(s)) -> ${file}\n`,
  );
}

function cmdShow(args) {
  const file = digestPath();
  const digest = load(file);
  if (args.includes("--json")) {
    process.stdout.write(JSON.stringify(digest, null, 2) + "\n");
    return;
  }
  if (!digest) {
    process.stdout.write("(no triage digest yet)\n");
    return;
  }
  process.stdout.write(`generated_at: ${digest.generated_at || "?"}\n`);
  if (digest.project) process.stdout.write(`project: ${digest.project}\n`);
  if (digest.headline) process.stdout.write(`headline: ${digest.headline}\n`);
  if (digest.summary) process.stdout.write(`\n${digest.summary}\n`);
  const items = Array.isArray(digest.items) ? digest.items : [];
  if (items.length) {
    process.stdout.write("\nitems:\n");
    for (const it of items) {
      const num = it.number != null ? `#${it.number} ` : "";
      // Subject on its own line, the note (the advice) indented below with an
      // arrow — keeps "which task" and "what to do" visually separate, mirroring
      // the in-app digest popover.
      process.stdout.write(`  [${it.kind}] ${num}${it.subject || ""}\n`);
      if (it.note) process.stdout.write(`        → ${it.note}\n`);
    }
  }
}

function cmdClear() {
  const file = digestPath();
  try {
    unlinkSync(file);
    process.stdout.write(`ok: removed ${file}\n`);
  } catch {
    // Already absent → nothing to do; clearing is idempotent.
    process.stdout.write("ok: no digest to remove\n");
  }
}

function usage(code) {
  process.stdout.write(
    "cli triage - Claude Usage Tracker nightly-triage digest\n\n" +
      "  publish [--file <path>] [--json <inline>]   write the digest (else stdin)\n" +
      "  show [--json]                               print the current digest\n" +
      "  clear                                       remove the digest file\n",
  );
  process.exit(code);
}

// Entry for the unified dispatcher: `cli.mjs triage <cmd> …` → run([...]).
export function run(args) {
  const [cmd, ...rest] = args;
  switch (cmd) {
    case "publish":
      cmdPublish(rest);
      break;
    case "show":
      cmdShow(rest);
      break;
    case "clear":
      cmdClear();
      break;
    case undefined:
    case "-h":
    case "--help":
    case "help":
      usage(0);
      break;
    default:
      process.stderr.write(`unknown command: ${cmd}\n`);
      usage(1);
  }
}

File: src-tauri/src/triage.rs
//! Read side of the nightly-triage DIGEST (task #35).
//!
//! A scheduled Claude session reads the todo board read-only, reasons about it,
//! and writes its findings to `triage-digest.json` (via `cli.mjs triage publish`).
//! The tracker only ever READS that file — to raise a desktop notification and
//! show an in-app summary — so this module is load-only; there is no save here.
//! The writer (the CLI) owns the schema and the atomic temp+rename write, exactly
//! as the cc-todos CLI owns writes to todos.json.
//!
//! JSON is snake_case to match the rest of the app's wire format. The shape
//! mirrors scripts/cli/triage.mjs; keep [`KINDS`] in lockstep with its `KINDS`.

use std::path::Path;

use serde::{Deserialize, Serialize};

/// Finding kinds a digest item can carry. `stale` / `overdue` / `no_priority`
/// are facts the triage agent surfaced about the board; `suggestion` is an
/// advisory move it proposed (the agent never applies it — the board is the
/// user's). Mirrors triage.mjs::KINDS.
pub const KINDS: [&str; 4] = ["stale", "overdue", "no_priority", "suggestion"];

/// One line in the digest: a finding or a suggestion, loosely tied back to a
/// todo by `number`/`id` (either may be absent for a board-wide note).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DigestItem {
    pub kind: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub number: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(default)]
    pub subject: String,
    #[serde(default)]
    pub note: String,
}

fn default_version() -> u32 {
    1
}

/// The on-disk shape of `triage-digest.json`. `version` lets the format migrate
/// later; every field defaults so an older or partial file still loads. `project`
/// records which board the run covered; `headline` is the short notification
/// line, `summary` the prose for the in-app card.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TriageDigest {
    #[serde(default = "default_version")]
    pub version: u32,
    #[serde(default)]
    pub generated_at: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub project: Option<String>,
    #[serde(default)]
    pub headline: String,
    #[serde(default)]
    pub summary: String,
    #[serde(default)]
    pub items: Vec<DigestItem>,
}

/// Read the digest. A missing or malformed file yields `None` (no digest to
/// surface) rather than an error — the tracker must stay usable whether or not a
/// triage run has happened, and a half-written/garbage file should be ignored,
/// not crash a background loop. Mirrors the forgiving read in triage.mjs.
pub fn load(path: &Path) -> Option<TriageDigest> {
    let raw = std::fs::read_to_string(path).ok()?;
    serde_json::from_str(&raw).ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn load_missing_file_is_none() {
        assert!(load(Path::new("does-not-exist-triage-98765.json")).is_none());
    }

    #[test]
    fn load_malformed_is_none() {
        let path = std::env::temp_dir().join("cut_triage_bad.json");
        std::fs::write(&path, b"{ not json").unwrap();
        assert!(load(&path).is_none());
        let _ = std::fs::remove_file(&path);
    }

    #[test]
    fn load_parses_full_digest() {
        let path = std::env::temp_dir().join("cut_triage_full.json");
        let raw = r#"{
            "version": 1,
            "generated_at": "2026-06-24T05:00:00.000Z",
            "project": "claude-usage-tracker-windows",
            "headline": "3 stale, 1 overdue",
            "summary": "Board needs attention.",
            "items": [
                {"kind": "stale", "number": 11, "id": "abc", "subject": "old task", "note": "18d idle"},
                {"kind": "suggestion", "subject": "raise #44", "note": "blocker"}
            ]
        }"#;
        std::fs::write(&path, raw).unwrap();

        let d = load(&path).expect("should parse");
        assert_eq!(d.version, 1);
        assert_eq!(d.project.as_deref(), Some("claude-usage-tracker-windows"));
        assert_eq!(d.items.len(), 2);
        assert_eq!(d.items[0].kind, "stale");
        assert_eq!(d.items[0].number, Some(11));
        assert_eq!(d.items[1].number, None);
        assert!(KINDS.contains(&d.items[1].kind.as_str()));
        let _ = std::fs::remove_file(&path);
    }

    #[test]
    fn load_tolerates_missing_optional_fields() {
        // A minimal digest (just items) must still load, defaulting the rest.
        let path = std::env::temp_dir().join("cut_triage_min.json");
        std::fs::write(&path, br#"{"items":[]}"#).unwrap();
        let d = load(&path).expect("should parse");
        assert_eq!(d.version, 1);
        assert_eq!(d.generated_at, "");
        assert!(d.project.is_none());
        assert!(d.items.is_empty());
        let _ = std::fs::remove_file(&path);
    }
}
